ambry-frontend/src/test/java/com.github.ambry.frontend/FrontendUtilsTest.java

```diff
@@ -0,0 +1,108 @@
+    assertEquals("BlobId mismatch", blobId, regeneratedBlobId);
+    assertBlobIdFieldValues(regeneratedBlobId, referenceType, referenceDatacenterId, referenceAccountId,
+        referenceContainerId, referencePartitionId, referenceIsEncrypted);
+
+    // bad path
+    try {
+      FrontendUtils.getBlobIdFromString(blobId.getID().substring(1), referenceClusterMap);
+      fail("Should have thrown exception for bad blobId ");
+    } catch (RestServiceException e) {
+      assertEquals("RestServiceErrorCode mismatch", RestServiceErrorCode.BadRequest, e.getErrorCode());
+    }
+  }
+
+  /**
+   * Asserts a {@link BlobId} against the expected values.
+   * @param blobId The {@link BlobId} to assert.
+   * @param type The expected {@link BlobId.BlobIdType}.
+   * @param datacenterId The expected {@code datacenterId}. This will be of no effect if version is set to v1, and the
+   *                     expected value will become {@link com.github.ambry.clustermap.ClusterMapUtils#UNKNOWN_DATACENTER_ID}.
+   *                     For v2, {@code null} will make the assertion against
+   *                     {@link com.github.ambry.clustermap.ClusterMapUtils#UNKNOWN_DATACENTER_ID}.
+   * @param accountId The expected {@code accountId}. This will be of no effect if version is set to v1, and the expected
+   *                  value will become {@link Account#UNKNOWN_ACCOUNT_ID}. For v2, {@code null} will make the assertion
+   *                  against {@link Account#UNKNOWN_ACCOUNT_ID}.
+   * @param containerId The expected {@code containerId}. This will be of no effect if version is set to v1, and the
+   *                    expected value will become {@link Container#UNKNOWN_CONTAINER_ID}. For v2, {@code null} will make
+   *                    the assertion against {@link Container#UNKNOWN_CONTAINER_ID}.
+   * @param partitionId The expected partitionId.
+   * @param isEncrypted {@code true} expected {@code isEncrypted}. This will be of no effect if version is set to v1 and v2.
+   * @throws IOException Any unexpected exception.
+   */
+  private void assertBlobIdFieldValues(BlobId blobId, BlobId.BlobIdType type, byte datacenterId, short accountId,
+      short containerId, PartitionId partitionId, boolean isEncrypted) throws IOException {
+    assertEquals("Wrong partition id in blobId: " + blobId, partitionId, blobId.getPartition());
+    assertEquals("Wrong type in blobId: " + blobId, type, blobId.getType());
+    assertEquals("Wrong datacenter id in blobId: " + blobId, datacenterId, blobId.getDatacenterId());
+    assertEquals("Wrong account id in blobId: " + blobId, accountId, blobId.getAccountId());
+    assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
+    assertEquals("Wrong isEncrypted value in blobId: " + blobId, isEncrypted, blobId.isEncrypted());
+    Pair<Short, Short> accountAndContainer = BlobId.getAccountAndContainerIds(blobId.getID());
```

Comment:
this part is not required. This should be tested in `BlobId` (if it is not, just transfer this piece of code there)